helix-core/src/diff.rs

```diff
@@ -0,0 +1,72 @@
+use ropey::Rope;
+
+use crate::{Change, Transaction};
+
+/// Compares `old` and `new` to generate a [`Transaction`] describing
+/// the steps required to get from `old` to `new`.
+pub fn compare_ropes(old: &Rope, new: &Rope) -> Transaction {
+    use std::borrow::Cow;
+
+    // `similar` only works on contiguous data, so a `Rope` has
+    // to be temporarily converted into a vec until the diff
+    // is created.
+    let old_vec: Vec<_> = old.lines().map(Cow::from).collect();
+    let new_vec: Vec<_> = new.lines().map(Cow::from).collect();
```

Comment:
I thought we were going to do the diff on chars?  Not as in collecting to `Vec<char>` but as in `String`.

I mean, by line is fine for now.  It won't work well for files with especially long lines (e.g. minimized .js files, etc.).  But that isn't something to block the PR on, I think.  I guess I'm just curious why you changed direction.

Reply:
I believe @archseer wanted to do something other than chars

Reply:
For user-facing diffs (for e.g. side-by-side diff views) I definitely agree that line diffs are the way to go.  But in this case the diff is essentially a way of compressing file differences to store on the undo stack efficiently.  So I think the use-case here is a little different.  (But again, I don't want to block merging on this, either way.)

@archseer Thoughts?

Reply:
I would like to diff on chars, as long as it's done on `String` (which doesn't need allocations) and not `Vec<char>` (the collect() will do a ton of extra work to expand the string to chars )